vllm/spec_decode/batch_expansion.py

```diff
@@ -247,24 +249,36 @@ def _create_target_seq_group_metadata(
         token_ids_to_score = self._get_token_ids_to_score(
             proposal_token_ids[batch_index])
 
+        # Use simpler sampling parameters apart from for final token
+        # (in particular don't do seeded sampling) since those sampled tokens
+        # aren't used
```

Comment:
I don't completely follow this optimization. Do we really only use the sampled token in the case that all proposals are accepted? My understanding was that if we accept e.g., 2 proposed tokens, we can still get a third one by sampling after verification? Don't we need the sampling params to be correct in all cases?

Reply:
> Do we really only use the sampled token in the case that all proposals are accepted?

Yes, it's only used for the bonus token.

> My understanding was that if we accept e.g., 2 proposed tokens, we can still get a third one by sampling after verification?

Yes but in this case the rejection sampler uses a "recovered" distribution based on the target and draft logits, so it's essentially a separate sampling operation. Sampled tokens from the batch-expanded target forward aren't used.

This can be seen [here](https://github.com/vllm-project/vllm/blob/main/vllm/spec_decode/spec_decode_worker.py#L585) where just the final sampled token is sliced off from the contracted sequence.

Reply:
Right, I see now. The rejection sampler uses the recovered distribution if we reject one of the proposals, and indeed the only case where this doesn't happen is if *all* proposals are accepted (e.g. bonus token logic).